File: controlm-saas/1-general-examples/code-snippets/python/code_blocks.py
#!/bin/python3
# example Python code for starting a script to communiocate with Control-M AAPI.
# reads command-line arguments, then reads API Token and sends an example API request.
import requests
import json
import argparse
import os, sys
import getpass

# API URL - MODIFY THIS
ctmaapi_url = 'https://YOURTENANT-aapi.us1.controlm.com/automation-api' 

# Parse command line arguments
parser = argparse.ArgumentParser(description='Updates Authorized Control-M/Servers list in multiple Agents')
parser.add_argument('-tf', '--tokenfile', dest='tokenfile', type=str, help='File that contains the API Key', required=False)
args = parser.parse_args()


# Read API Token from file, if given. Else, request token from user interactively (without echoing input)
if args.tokenfile is not None:
    if not os.path.exists(args.tokenfile):
        print("Unable to read the specified file, %s, or file does not exist." % args.tokenfile)
    else:
        with open(args.tokenfile, "r") as f:
            apitoken = f.read()
            apitoken = apitoken.rstrip(' \t\r\n\0')
else:
    apitoken = getpass.getpass("API Token: ")

# Example get roles request (make sure the token has config authorizations)
try:
    r = requests.get( ctmaapi_url+'/config/authorization/roles',
        headers={"x-api-key" : apitoken})
except Exception as err: # Catch any general errors
    print("Connecting to Automation API REST Server failed with error: " + str(err))
    sys.exit(1)

resp=json.loads(r.text)
print("Get Roles Response:")
print(json.dumps(resp,indent=4))

File: helix-control-m/1-general-examples/code-snippets/perl/code_blocks.pl
#!/usr/bin/env perl
use JSON;
use JSON::Parse 'valid_json';
use REST::Client;
use IO::Socket::SSL qw( SSL_VERIFY_NONE );
use Getopt::Long;

## API URL - MODIFY THIS
$ctmapi_url='https://bmcl2-1-4188-aapi.qa.controlm.com/automation-api';

# Script usage text
$usagetext = "$0 [--tokenfile filename]\n";
$helptext  = $usagetext .
  "  --tokenfile:   File containing the API token\n" .
  "  --help:     Print this text\n";

# Function for prompting for API token
sub prompt_for_token {
    require Term::ReadKey;

    Term::ReadKey::ReadMode('noecho');
    print "API Token: ";
    my $token = Term::ReadKey::ReadLine(0);
    Term::ReadKey::ReadMode('restore');
    $token =~ s/\R\z//;
    print "\n";

    return $token;
}

# Parse command line options
GetOptions(
  'tokenfile=s' => \$tokenfile,
  'help'     => \$help
) or die $usagetext;

if($help) { die($helptext); } # if --help show help text and end

# if token file was specified, read token from file, else prompt user
if( $pwfile ) {
  if( -e $pwfile ) {
    open my $file, '<', $pwfile;
    $token = <$file>;
    chomp($token);
    close $file;
  }
  else {
    die("File $pwfile not found!")
  }
}
else {
  $token=prompt_for_token();
}

# initialize REST client
my $client = REST::Client->new();
$client->setHost($ctmapi_url);

# example Get Roles request (make sure the token has config authorizations)
$client->addHeader('Content-Type', 'application/json');
$client->addHeader('x-api-key', $token);
$client->GET('/config/authorization/roles', $reqbody);

# Handle errors
if ($client->responseCode() != 200) {
  if( valid_json($client->responseContent()) ) {
    $json=from_json($client->responseContent());
    print $json->{errors}[0]->{message} . "\n";
  }
  else {
    print $client->responseContent() . "\n";
 }
 exit(1);
}

#  print response
print "Get Roles Response:\n";
print $client->responseContent() . "\n";

